elfi/tools.py

```diff
@@ -0,0 +1,130 @@
+from functools import partial
+import numpy as np
+
+
+__all__ = ['vectorize']
+
+
+def run_vectorized(operation, *inputs, constants=None, batch_size=None, **kwargs):
+    """Runs the operation as if it was vectorized over the individual runs in the batch.
+
+    Helper for cases when you have an operation that does not support vector arguments.
+
+    Parameters
+    ----------
+    operation : callable
+        operation that will be run `batch_size` times.
+    inputs
+    constants : tuple or int, optional
+        indexes of arguments in inputs that are constants
+    batch_size : int, optional
+    kwargs
+
+    Notes
+    -----
+    This is an experimental feature.
+
+    This is a convenience method and uses a for loop for vectorization. For best
+    performance, one should aim to implement vectorized operations (by using e.g. numpy
+    functions that are mostly vectorized) if at all possible.
+
+    If the output from the operation is not a numpy array or if the shape of the output
+    in different runs differs, the `dtype` of the returned numpy array will be `object`.
+
+    Returns
+    -------
+    operation_output
+        If batch_size > 1, a numpy array of outputs is returned
+    """
+
+    constants = [] if constants is None else list(constants)
+
+    # Check input and set constants and batch_size if needed
+    for i, inpt in enumerate(inputs):
+        if i in constants:
+            continue
+
+        try:
+            l = len(inpt)
+        except:
+            constants.append(i)
+            l = 1
+
+        if l != 1:
+            if batch_size is None:
+                batch_size = l
+            elif batch_size != l:
```

Comment:
The batch size would come from the framework. But not all nodes receive it as input, so in that case we try to infer it.